### ISSUE TITLE
Using tree reductions with unique

### ISSUE THREAD

**Issue by jakirkham** [dask/dask#2851]
The current `unique` implementation, while lazy, does not currently leverage tree reductions akin to those used for things like `sum`. These tree reductions allow operations to be computed over each chunk and then reduce small groups of chunks into a smaller group of new chunks that can be further reduced until only one small chunk remains. Through this process it is possible to work with much larger pieces of data. To allow `unique` to work on larger pieces of data, it would be nice to leverage this tree reduction strategy in its implementation.

It is almost possible to implement `unique` using the existing reduction utilities in [`dask.array.reductions`]( https://github.com/dask/dask/blob/b9e3487a963b574f91eb403fdfc0baf1bc97d2eb/dask/array/reductions.py ) except that these functions seem to assume that a reduction results in a chunk with a single element (or one element for a dimension). However `unique` may return a unknown number of elements that could range from one to the length of the data itself. Hence we would need a way to explain that the chunks from each reduction may have a different or indeterminate size.

**Comment by jakirkham** [comment]
Potentially this can benefit from the strategy discussed in issue ( https://github.com/dask/dask/issues/2899 ).

**Comment by convexset** [comment]
Might be interesting to look at.

Either it should be using `reduction` or recursion.

With recursion, there have to be some norms about how to go about combining the intermediates. That is, typically how many unique elements are to be expected in each chunk. Because this has to be considered a priori.

In fact, it might be good to extend this into a ["counter"](https://docs.python.org/3/library/collections.html#collections.Counter) type method, and `unique` would use the output and just extract the "keys".

**Comment by jakirkham** [comment]
If you're interested, would take a look at [`dask.array.reductions`]( https://github.com/dask/dask/blob/b9e3487a963b574f91eb403fdfc0baf1bc97d2eb/dask/array/reductions.py ).

The way the [`unique` kernel function]( https://github.com/dask/dask/blob/0.18.0/dask/array/routines.py#L745 ) is written should be friendly to reductions. Happy to answer questions about it.

Interesting suggestion with the `counter` method. Though I don't completely follow what you are thinking here. Could you please explain a bit more what you have in mind?

**Comment by convexset** [comment]
Oh. I see a `return_counts` argument. I previously only did a little plot of the compute graph and didn't look at the code/docs.

**Comment by jakirkham** [comment]
Yeah it supports all the arguments that NumPy's `unique` supports except `axis`. Also made a small update to the comment above as it inaccurately referenced `unique` as the kernel function, which is instead `_unique_internal`. Hopefully the `_unique_internal` doesn't need to change at all, but `unique` would need to change to use tree reductions with `_unique_internal`.

**Comment by jakirkham** [comment]
Just as an observation, if one wanted to tackle this, looking at the `arg_*` functions in `dask.array.reductions` would likely be insightful. That case much like `unique` has to pass around a few different parts of the result (e.g. max values and their indices per chunk) and makes the same choice to leverage structured arrays to make this happen. It's not totally one-to-one as `unique` can return more than a single value and there are multiple (optional) pieces to `unique`'s result.
